src/Psalm/Internal/Analyzer/Statements/Expression/CastAnalyzer.php

```diff
@@ -183,22 +183,44 @@ public static function analyze(
         }
 
         if ($stmt instanceof PhpParser\Node\Expr\Cast\Object_) {
-            $was_inside_general_use = $context->inside_general_use;
-            $context->inside_general_use = true;
-            if (ExpressionAnalyzer::analyze($statements_analyzer, $stmt->expr, $context) === false) {
-                $context->inside_general_use = $was_inside_general_use;
+            $result = self::checkExprGeneralUse($statements_analyzer, $stmt, $context);
 
+            if (!$result) {
                 return false;
             }
-            $context->inside_general_use = $was_inside_general_use;
 
-            $type = new Union([new TNamedObject('stdClass')]);
+            $permissible_atomic_types = [];
+            $all_permissible = false;
 
-            $maybe_type = $statements_analyzer->node_data->getType($stmt->expr);
+            if ($stmt_expr_type = $statements_analyzer->node_data->getType($stmt->expr)) {
+                if ($stmt_expr_type->isObjectType()) {
+                    self::handleRedundantCast($stmt_expr_type, $statements_analyzer, $stmt);
+                }
+
+                $all_permissible = true;
+
+                foreach ($stmt_expr_type->getAtomicTypes() as $type) {
+                    if ($type instanceof Scalar) {
+                        $objWithProps = new TObjectWithProperties(['scalar' => new Union([$type])]);
```

Comment:
It's a nice touch to handle this, but do we really want to?

This is a very weird behaviour by PHP and, IMHO, rarely intended by any developper.

We could throw a InvalidCast or a PossiblyInvalidCast when encountering a scalar in an object cast, what do you think?

Reply:
It's definitely weird behavior by PHP, and probably rarely intended by devs. But I still feel like Psalm should infer the correct type when possible rather than throwing an error, especially when it's easy to do so. Scalars cast to an array are also quite uncommon, but Psalm does correctly infer the type for these.